File: app/src/main/java/com/challenge/fidoreader/fagment/ReaderButtonFragment.java
package com.challenge.fidoreader.fagment;

import android.content.Context;
import android.content.DialogInterface;
import android.graphics.Color;
import android.graphics.Typeface;
import android.os.AsyncTask;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;
import android.widget.Toast;

import com.challenge.fidoreader.Exception.UserException;
import com.challenge.fidoreader.MainActivity;
import com.challenge.fidoreader.R;
import com.challenge.fidoreader.Util.Util;
import com.challenge.fidoreader.fidoReader.Authenticator;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.cbor.CBORFactory;
import com.fasterxml.jackson.dataformat.cbor.CBORParser;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.Map;

public class ReaderButtonFragment extends Fragment {
    public final static String TAG = "ReaderButtonFragment";

    private ProgressBar pgsBar;

    ImageView imageView;
    TextView txtView2;
    TextView txtView3;

    Button getCredListbtn;
    Button getInfobtn;
    Button clientpinbtn;
    Button enrollManageBtn;
    Button resetBtn;

    GetInfoResponseBottomSheetDialog getinfoSheet;
    Map<String, Object> getInfoResponse;

    MainActivity mainActivity;
    private ReaderButtonViewModel mViewModel;

    public static ReaderButtonFragment newInstance() {
        return new ReaderButtonFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState){
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater, @Nullable final ViewGroup container, @Nullable final Bundle savedInstanceState) {
        Log.v(TAG, "onCreateView");

        final View view = inflater.inflate(R.layout.fragment_reader_button, container, false);

        getCredListbtn = (Button)view.findViewById(R.id.readerActivationBtn);
        getCredListbtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Log.v(TAG, "onClick");
                try{
                    InputPINFragment inputPINFragment = new InputPINFragment(mainActivity, "Credential");
                    inputPINFragment.show(getFragmentManager(), "dialog");
                    inputPINFragment.setCancelable(false);
                    inputPINFragment.setDialogResult(new InputPINFragment.OnDialogResult() {
                        @Override
                        public void finish(String result) {
                            if(result.equals("OK")){
                                GoogleTranslate googleTranslate = new GoogleTranslate();
                                googleTranslate.execute();
                            }
                        }
                    });


                }catch (Exception e){
                    e.printStackTrace();
                }
            }
        });
        getCredListbtn.setEnabled(false);

        getInfobtn = (Button)view.findViewById(R.id.KonaBIOPASSGetInfoBtn);
        clientpinbtn = (Button)view.findViewById(R.id.KonaBIOPASSPINBtn);
        clientpinbtn.setEnabled(false);

        getInfobtn.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View v) {
                getinfoSheet = new GetInfoResponseBottomSheetDialog(clientpinbtn, mainActivity, getInfoResponse);
                getinfoSheet.show(getActivity().getSupportFragmentManager(), "exampleBottomSheet");

            }
        });
        // getInfobtn.setEnabled(false);

        clientpinbtn.setOnClickListener(new View.OnClickListener(){

            @Override
            public void onClick(View v) {
                // Change PIN or
                // Set New PIN
                DialogFragment pinFragment = null;
                if(clientpinbtn.getText().toString().equals("Set New PIN")) {
                    pinFragment = new SetNewPINFragment(mainActivity);
                }
                else if(clientpinbtn.getText().toString().equals("Change PIN")) {
                    pinFragment = new ChangePINFragment(mainActivity);
                }

                pinFragment.show(getFragmentManager(), "dialog");
                pinFragment.setCancelable(false);
                try {
                    mainActivity.authenticator.setTag(mainActivity.cardReader.myTag);
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        });

        imageView = (ImageView)view.findViewById(R.id.readerActivationImageView);
        txtView2 = (TextView)view.findViewById(R.id.readerActivationText2);
        txtView3 = (TextView)view.findViewById(R.id.readerActivationText3);

        enrollManageBtn = view.findViewById(R.id.EnrollBtn);

        enrollManageBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Log.v(TAG, "onClick");
                try{
                    InputPINFragment inputPINFragment = new InputPINFragment(mainActivity, "Fingerprint");
                    inputPINFragment.show(getFragmentManager(), "dialog");
                    inputPINFragment.setCancelable(false);

                    inputPINFragment.setDialogResult(new InputPINFragment.OnDialogResult() {
                        @Override
                        public void finish(String result) {
                            if(result.equals("OK")){
                                GetEnrollInformation googleTranslate = new GetEnrollInformation();
                                googleTranslate.execute();
                            }
                        }
                    });


                }catch (Exception e){
                    e.printStackTrace();
                }
            }
        });

        resetBtn = view.findViewById(R.id.resetBtn);

        resetBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Log.v(TAG, "resetBtn");
                try{
                    ResetProcess googleTranslate = new ResetProcess();
                    googleTranslate.execute();

                }catch (Exception e){
                    e.printStackTrace();
                }
            }
        });

        pgsBar = (ProgressBar) view.findViewById(R.id.h_progressbar);
        //pgsBar.setIndeterminate(true);
        pgsBar.setVisibility(View.GONE);

        try {
            setResult(MainActivity.cardReader.result_image,
                    MainActivity.cardReader.result1_str,
                    MainActivity.cardReader.result2_str,
                    MainActivity.cardReader.result);
        } catch (Exception e) {
            e.printStackTrace();
        }

        return view;
    }

    public boolean isReady(){
        if(imageView != null && txtView2 != null && txtView3 != null){
            Log.v(TAG, "Fragment is ready");
            return true;
        }
        Log.v(TAG, "Fragment is not ready");
        return false;
    }


    public void setEnabled() throws Exception {
        Log.v(TAG, "setEnabled");
        getCredListbtn.setEnabled(true);

        getInfobtn.setEnabled(true);
        getInfoResponse = getInfo();
        clientpinbtn.setEnabled(true);
    }
    public void setResult(int resource, String str1, String str2){
        imageView.setImageResource(resource);
        txtView2.setText(str1);
        txtView3.setText(str2);
    }


    public void setResult(int resource, String str1, String str2, boolean result) throws Exception {
        setResult(resource, str1, str2);
        if(result){
            setEnabled();
        }
    }


    @Override
    public void onResume(){
        super.onResume();
        Log.v(TAG, "onResume");
        setResult(MainActivity.cardReader.result_image, MainActivity.cardReader.result1_str, MainActivity.cardReader.result2_str);
    }

    @Override
    public void onPause(){
        super.onPause();
        Log.v(TAG, "onPause");

    }


    @Override
    public void onStop(){
        super.onStop();
        Log.v(TAG, "onStop");
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        mainActivity = (MainActivity)getActivity();
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mainActivity = null;
        Log.v(TAG, "onDetach");
    }


    @Override
    public void onDestroyView(){
        super.onDestroyView();
        Log.v(TAG, "onDestroyView");
    }

    @Override
    public void onDestroy(){
        super.onDestroy();
        Log.v(TAG, "onDestroy");
    }

    class GoogleTranslate extends AsyncTask<Object, Object, Object> {
        ArrayList<CredentialItem> list = null;

        @Override
        protected Object doInBackground(Object... params) {
            Log.v("translate", "doInBackground");

            try {
                Authenticator authenticator = mainActivity.authenticator;
                authenticator.setTag(mainActivity.cardReader.myTag);
                list = authenticator.getCredentialList();
            }
            catch (Exception e) {
                if(e.getMessage().equals("2E")){
                    list = new ArrayList<CredentialItem>();
                }
                e.printStackTrace();
            }
            return list;
        }

        @Override
        protected void onProgressUpdate(Object... values) {
            super.onProgressUpdate(values);
            Log.v("translate", "\n"+values[0].toString()+"번 count했습니다.");
        }


        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            Log.v("translate", "onPreExecute");
            pgsBar.setVisibility(View.VISIBLE);
        }

        @Override
        protected void onPostExecute(Object s) {
            super.onPostExecute(s);
            Log.v("translate", "onPostExecute");
            pgsBar.setVisibility(View.GONE);
            mainActivity.onChangeFragmentToList(list);
        }

    }

    class GetEnrollInformation extends AsyncTask<Object, Object, Object> {

        ArrayList<FingerItem> list = null;
        @Override
        protected Object doInBackground(Object... params) {
            Log.v("GetEnrollInformation", "doInBackground");
            try {
                Authenticator authenticator = mainActivity.authenticator;
                authenticator.setTag(mainActivity.cardReader.myTag);
                list = authenticator.readEnrollInformation();
            }
            catch (Exception e) {
                Log.v("GetEnrollInformation", "" + e.getMessage());
                if(e.getMessage().equals("2C")){
                    list = new ArrayList<FingerItem>();
                }
                e.printStackTrace();
            }
            return list;
        }

        @Override
        protected void onProgressUpdate(Object... values) {
            super.onProgressUpdate(values);
            Log.v("translate", "\n"+values[0].toString()+"번 count했습니다.");
        }


        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            Log.v("translate", "onPreExecute");
            pgsBar.setVisibility(View.VISIBLE);
        }

        @Override
        protected void onPostExecute(Object s) {
            super.onPostExecute(s);
            Log.v("translate", "onPostExecute");
            pgsBar.setVisibility(View.GONE);
            mainActivity.onChangeFragmentToList2(list);

        }

    }


    class ResetProcess extends AsyncTask<Object, Object, Object> {

        boolean result = true;
        @Override
        protected Object doInBackground(Object... params) {
            Log.v("ResetProcess", "doInBackground");
            try {
                Authenticator authenticator = mainActivity.authenticator;
                authenticator.setTag(mainActivity.cardReader.myTag);
                result = authenticator.reset();
            }
            catch (Exception e) {
                e.printStackTrace();
            }
            return result;
        }

        @Override
        protected void onProgressUpdate(Object... values) {
            super.onProgressUpdate(values);
        }


        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            Log.v("ResetProcess", "onPreExecute");
            pgsBar.setVisibility(View.VISIBLE);
        }

        @Override
        protected void onPostExecute(Object s) {
            super.onPostExecute(s);
            Log.v("ResetProcess", "onPostExecute");
            pgsBar.setVisibility(View.GONE);
            Toast.makeText(mainActivity,"정상종료", Toast.LENGTH_SHORT).show();

        }

    }



    private Map<String, Object> getInfo() throws Exception {
        String result = null;

        mainActivity.authenticator.setTag(mainActivity.cardReader.myTag);
        result = mainActivity.authenticator.getInfo();

        ByteArrayInputStream bais = new ByteArrayInputStream(Util.atohex(result));

        CBORFactory cf = new CBORFactory();
        ObjectMapper mapper = new ObjectMapper(cf);

        CBORParser cborParser = null;
        cborParser = cf.createParser(bais);
        Map<String, Object> responseMap = null;
        responseMap = mapper.readValue(cborParser, new TypeReference<Map<String, Object>>() {
        });

        for (String key : responseMap.keySet()) {
            if (key.equals("4")) {
                LinkedHashMap<String, Boolean> options = (LinkedHashMap<String, Boolean>) responseMap.get(key);

                if(options.get("clientPin")){
                    clientpinbtn.setText("Change PIN");
                }
                else{
                    clientpinbtn.setText("Set New PIN");
                }
            }
        }

        return responseMap;
    }
}

File: app/src/main/java/com/challenge/fidoreader/frag/reader/ReaderButtonFragment.kt
package com.challenge.fidoreader.frag.reader

import android.content.Context
import android.os.AsyncTask
import android.os.Bundle
import android.os.Parcelable
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.fragment.app.Fragment
import com.challenge.fidoreader.*
import com.challenge.fidoreader.Util.ParcelableActivityData
import com.challenge.fidoreader.Util.atohex
import com.challenge.fidoreader.Util.cardReader
import com.challenge.fidoreader.frag.reader.InputPINFragment.OnDialogResult
import com.challenge.fidoreader.fidoReader.Authenticator
import com.challenge.fidoreader.fidoReader.data.CBOR_CODE_CTAP2_ERR_INVALID_OPTION
import com.challenge.fidoreader.fidoReader.data.CBOR_CODE_CTAP2_ERR_NO_CREDENTIALS
import com.challenge.fidoreader.reader.CardReader
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.cbor.CBORFactory
import com.fasterxml.jackson.dataformat.cbor.CBORParser
import java.io.ByteArrayInputStream
import java.util.*
import kotlin.collections.ArrayList


class ReaderButtonFragment : Fragment(){
    val TAG = "ReaderButtonFragment"

    lateinit var imageView: ImageView
    lateinit var textView2: TextView
    lateinit var textView3: TextView

    lateinit var getCredListbtn :Button
    lateinit var getinfoBtn:Button
    lateinit var clientPinBtn:Button
    lateinit var enrollManageBtn:Button
    lateinit var resetBtn:Button

    var hasClientPIN = false

    lateinit var mainActivity:MainActivity

    var pgsBar: ProgressBar? = null

    var getinfoSheet: GetInfoResponseBottomSheetDialog? = null
    var getInfoResponse: Map<String, Any>? = null

    var isReady = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        Log.v(TAG, "onCreateView")

        var mainView: View = inflater.inflate(R.layout.fragment_reader_button, container, false)

        getCredListbtn  = mainView.findViewById(R.id.readerActivationBtn)
        imageView = mainView.findViewById(R.id.readerActivationImageView)
        textView2 = mainView.findViewById(R.id.readerActivationText2)
        textView3 = mainView.findViewById(R.id.readerActivationText3)
        enrollManageBtn = mainView.findViewById(R.id.EnrollBtn)
        resetBtn = mainView.findViewById(R.id.resetBtn)
        pgsBar = mainView.findViewById(R.id.h_progressbar)
        getinfoBtn = mainView.findViewById(R.id.KonaBIOPASSGetInfoBtn)
        clientPinBtn = mainView.findViewById(R.id.KonaBIOPASSPINBtn)

        getCredListbtn .setOnClickListener{
            Log.v(TAG, "onClick")
            try {
                val inputPINFragment = InputPINFragment(mainActivity, "Credential")
                inputPINFragment.show(fragmentManager!!, "dialog")
                inputPINFragment.isCancelable = false
                inputPINFragment.setDialogResult(object : OnDialogResult {
                    override fun finish(result: String?) {
                        Log.v(TAG, "inputPINFragment result : " + result)

                        if (result == "OK") {
                            val googleTranslate = GoogleTranslate()
                            googleTranslate.execute()
                        }
                    }
                })
            } catch (e: java.lang.Exception) {
                e.printStackTrace()
            }

        }
        getCredListbtn .isEnabled = false


        getinfoBtn.setOnClickListener{
            getinfoSheet = GetInfoResponseBottomSheetDialog(clientPinBtn, mainActivity, getInfoResponse!!)
            getinfoSheet!!.show(activity!!.supportFragmentManager, "exampleBottomSheet")
        }
        //getinfoBtn.isEnabled = false

        clientPinBtn.setOnClickListener {
            var pinFragment = if(clientPinBtn.text.toString()=="Set New PIN"){
                SetNewPINFragment(mainActivity)
            }else{
                ChangePINFragment(mainActivity)
            }

            pinFragment.show(fragmentManager!!, "dialog")
            pinFragment.isCancelable = false

        }
        clientPinBtn.isEnabled = false

        enrollManageBtn.setOnClickListener {
            Log.v(TAG, "onClick")
            try {
                val inputPINFragment = InputPINFragment(mainActivity, "Fingerprint")
                inputPINFragment.show(fragmentManager!!, "dialog")
                inputPINFragment.isCancelable = false

                inputPINFragment.setDialogResult(object : OnDialogResult {
                    override fun finish(result: String?) {
                        if (result == "OK") {
                            val googleTranslate = GetEnrollInformation()
                            googleTranslate.execute()
                        }
                    }
                })
            } catch (e: java.lang.Exception) {
                e.printStackTrace()
            }
        }

        resetBtn.setOnClickListener {
            Log.v(TAG, "resetBtn")
            try {
                val googleTranslate = ResetProcess()
                googleTranslate.execute()
            } catch (e: java.lang.Exception) {
                e.printStackTrace()
            }
        }

        pgsBar!!.visibility = View.GONE

        try {
            setResult(cardReader, cardReader.result)
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }

        isReady = true

        return mainView
    }

    fun setResult(c:CardReader){
        when(c.cntionState){
            CardReader.ConnectState.ICC_ON_ATR ->{
                imageView.setImageResource(R.drawable.ic_icc_on_atr)
            } CardReader.ConnectState.ICC_ON ->{
                imageView.setImageResource(R.drawable.ic_icc_on)
            } CardReader.ConnectState.ICC_OFF ->{
                imageView.setImageResource(R.drawable.ic_icc_off)
            }
        }

        when(c.nfcReaderState){
            CardReader.NfcReaderState.NO_READER ->{
                textView2.text = "NFC ENABLED"
            } CardReader.NfcReaderState.READER_ABLE ->{
                textView2.text = "NO READER DETECTED"
            }
        }

        textView3.text = c.cntionResult
    }

    fun setResult(c:CardReader, result: Boolean){
        setResult(c)
        if(result){
            setEnable()
        }
    }

    fun setEnable(){
        Log.v(TAG, "SetEnable")
        getCredListbtn.isEnabled = true

        //getInfo()
        getinfoBtn.isEnabled = true
        getInfoResponse = getInfo();
        clientPinBtn.isEnabled = true
    }

    private fun getInfo(): Map<String, Any>? {
        //MainActivity.authenticator.myTag = CardReader.myTag
        var result = MainActivity.authenticator.getInfo()
        val bais = ByteArrayInputStream(result.atohex())
        val cf = CBORFactory()
        val mapper = ObjectMapper(cf)
        var cborParser: CBORParser? = null
        cborParser = cf.createParser(bais)
        var responseMap: Map<String, Any>? = null
        responseMap = mapper.readValue(cborParser, object : TypeReference<Map<String?, Any?>?>() {})
        for (key in responseMap!!.keys) {
            if (key == "4") {
                val options = responseMap[key] as LinkedHashMap<String, Boolean>?
                if (options!!["clientPin"]!!) {
                    clientPinBtn.setText("Change PIN")
                } else {
                    clientPinBtn.setText("Set New PIN")
                }
            }
        }
        return responseMap
    }

    override fun onResume() {
        super.onResume()
        Log.v(TAG, "onResume")
        setResult(cardReader)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        mainActivity = (activity as MainActivity?)!!
    }

    override fun onDetach() {
        super.onDetach()
        //mainActivity = null
        Log.v(TAG, "onDetach")
    }

    abstract inner class AsyncTaskProcess  : AsyncTask<Any?, Any?, Any?>() {
        open var TAG = "AsyncTaskProcess"
        var list : ArrayList<Parcelable>? = null
        lateinit var errorcode :String
        lateinit var parcelableActivityData : ParcelableActivityData

        abstract fun process(authenticator: Authenticator) : ArrayList<Parcelable>

        override fun doInBackground(vararg params: Any?): Any? {
            Log.v(TAG, "doInBackground")
            try {
                val authenticator: Authenticator = MainActivity.authenticator
                //authenticator.myTag = CardReader.myTag
                list = process(authenticator)
            } catch (e: java.lang.Exception) {
                Log.d("AsyncTaskProcess : ", "Error : " + e.message)
                if (e.message.equals(errorcode)) {
                    list = ArrayList()
                }
                e.printStackTrace()
            }
            return list
        }

        override fun onProgressUpdate(vararg values: Any?) {
            super.onProgressUpdate(*values)
            Log.v(TAG, "${values[0]}번 count했습니다.".trimIndent())
        }

        override fun onPreExecute() {
            super.onPreExecute()
            Log.v(TAG, "onPreExecute")
            pgsBar!!.visibility = View.VISIBLE
        }

        override fun onPostExecute(s: Any?) {
            super.onPostExecute(s)
            Log.v(TAG, "onPostExecute")
            pgsBar!!.visibility = View.GONE
            mainActivity.showActivityList(parcelableActivityData, list)
        }
    }

    inner class GoogleTranslate : AsyncTaskProcess() {
        override var TAG = "GoogleTranslate"

        init {
            errorcode = CBOR_CODE_CTAP2_ERR_NO_CREDENTIALS
            parcelableActivityData =  ParcelableActivityData(
                    CredListActivity::class.java as Class<Any>,
                    true,
                    "Credential is not exist",
                    "Credentiallist"
            )
        }

        override fun process(authenticator: Authenticator): ArrayList<Parcelable> {
            return authenticator.getCredentialList() as ArrayList<Parcelable>
        }
    }

    inner class GetEnrollInformation : AsyncTaskProcess() {
        override var TAG = "GetEnrollInformation"

        init {
            errorcode = CBOR_CODE_CTAP2_ERR_INVALID_OPTION
            parcelableActivityData = ParcelableActivityData(
                    EnrollManageActivty::class.java as Class<Any>,
                    false,
                    "Enrollment is not exist",
                    "fingerItem"
            )
        }

        override fun process(authenticator: Authenticator): ArrayList<Parcelable> {
            return authenticator.readEnrollInformation() as ArrayList<Parcelable>
        }
    }

    inner class ResetProcess : AsyncTask<Any?, Any?, Any>() {
        var result = true
        override fun doInBackground(vararg params: Any?): Any? {
            Log.v("ResetProcess", "doInBackground")
            try {
                val authenticator: Authenticator = MainActivity.authenticator
                //authenticator.myTag = CardReader.myTag
                result = authenticator.reset()
            } catch (e: java.lang.Exception) {
                e.printStackTrace()
            }
            return result
        }

        override fun onProgressUpdate(vararg values: Any?) {
            super.onProgressUpdate(*values)
        }

        override fun onPreExecute() {
            super.onPreExecute()
            Log.v("ResetProcess", "onPreExecute")
            pgsBar!!.visibility = View.VISIBLE
        }

        override fun onPostExecute(s: Any) {
            super.onPostExecute(s)
            Log.v("ResetProcess", "onPostExecute")
            pgsBar!!.visibility = View.GONE
            Toast.makeText(mainActivity, "정상종료", Toast.LENGTH_SHORT).show()
        }
    }


}



